src/main/scala/analysis/Analysis.scala

```diff
@@ -500,206 +495,149 @@ trait MemoryRegionAnalysisMisc:
 
   private val mallocVariable = Register("R0", BitVecType(64))
 
-  private val loopEscapeSet: mutable.Set[CfgNode] = mutable.Set.empty
-
-  def loopEscape(n: CfgNode): Boolean = {
-    if (loopEscapeSet.contains(n)) {
-      return true
-    }
-    loopEscapeSet.add(n)
-    false
-  }
-
-  /** Find decl of variables from node predecessors */
-  def findDecl(variable: Variable, n: CfgNode): mutable.ListBuffer[CfgNode] = {
-    val decls: mutable.ListBuffer[CfgNode] = mutable.ListBuffer.empty
-    // if we have a temporary variable then ignore it
-    variable match {
-      case r: Register =>
-        for (pred <- n.pred) {
-          if (loopEscape(pred)) {
-            return mutable.ListBuffer.empty
+  /** Default implementation of eval.
+    */
+  def eval(exp: Expr, env: lattice.sublattice.Element, n: CfgNode): lattice.sublattice.Element = {
+    Logger.debug(s"evaluating $exp")
+    Logger.debug(s"env: $env")
+    Logger.debug(s"n: $n")
+    exp match {
+      case binOp: BinaryExpr =>
+        if (binOp.arg1 == stackPointer) {
+          val rhs: Expr = evaluateExpression(binOp.arg2, n, constantProp)
+          Set(StackRegion(getNextStackCount(), binOp.arg2, None))
+        } else {
+          val evaluation: Expr = evaluateExpression(binOp, n, constantProp)
+          if (evaluation.equals(binOp)) {
+            return env
           }
-          pred match {
-            case cmd: CfgCommandNode =>
-              cmd.data match {
-                case localAssign: LocalAssign =>
-                  if (localAssign.lhs == variable) {
-                    decls.addOne(pred)
-                  } else {
-                    decls.addAll(findDecl(variable, pred))
-                  }
-                case _ =>
-              }
-            case _ =>
+          eval(evaluation, env, n)
+        }
+      case bitVecLiteral: BitVecLiteral =>
+        if (globals.contains(bitVecLiteral.value)) {
+          val globalName = globals(bitVecLiteral.value)
+          Set(DataRegion(globalName, bitVecLiteral, None))
+        } else if (subroutines.contains(bitVecLiteral.value)) {
+          val subroutineName = subroutines(bitVecLiteral.value)
+          Set(DataRegion(subroutineName, bitVecLiteral, None))
+        } else if (globalOffsets.contains(bitVecLiteral.value)) {
+          val val1 = globalOffsets(bitVecLiteral.value)
+          if (subroutines.contains(val1)) {
+            val globalName = subroutines(val1)
+            Set(DataRegion(globalName, bitVecLiteral, None))
+          } else {
+            Set(DataRegion(s"Unknown_${bitVecLiteral}", bitVecLiteral, None))
           }
+        } else {
+          //throw new Exception(s"Unknown type for $bitVecLiteral")
+          // unknown region here
+          Set(DataRegion(s"Unknown_${bitVecLiteral}", bitVecLiteral, None))
+        }
+      case variable: Variable =>
+        if (variable.name.contains("#") || variable.equals(stackPointer)) {
+          return env
```

Comment:
rather than using `variable.name.contains("#")` we should match the `Register` and `LocalVar` types as relevant here

Reply:
Resolved. [Upcoming merge from yousif-memory-region-analysis]